### PR TITLE
expression: implement wrap Expression with cast functions

### PR DESCRIPTION
**Pull request by XuHuaiyu**
1. This commit **_splits_** castFunctionClass into 
castAsIntFunctionClass, castAsRealFunctionClass, castAsDecimalFunctionClass, castAsStringFunctionClass, castAsTimeFunctionClass, castAsDurationFunctionClass

2. This commits implements wrapping Expression with cast functions:
WrapWithCastAsInt, WrapWithCastAsReal, WrapWithCastAsDecimal, WrapWithCastAsString, WrapWithCastAsTime, WrapWithCastAsDuration

PTAL @shenli @hanfei1991

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -76,6 +78,192 @@ func (b *builtinCastSig) eval(row []types.Datum) (d types.Datum, err error) {
...
+		sig = &builtinCastDecimalAsIntSig{bf}
+	case types.ClassString:
+		switch args[0].GetType().Tp {
+		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
+			sig = &builtinCastTimeAsIntSig{bf}
+		case mysql.TypeDuration:
+			sig = &builtinCastDurationAsIntSig{bf}
+		default:
+			sig = &builtinCastStringAsIntSig{bf}
+		}
```
Comment (shenli) [inline expression/builtin_cast.go]: Miss Duration?

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -76,6 +78,192 @@ func (b *builtinCastSig) eval(row []types.Datum) (d types.Datum, err error) {
 	return d, errors.Errorf("unknown cast type - %v", b.tp)
 }
 
+type castAsIntFunctionClass struct {
```
Comment (shenli) [inline expression/builtin_cast.go]: Add some comment for the following types. For example:
We implement castAsXXFunctionClass for .....

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -558,3 +767,119 @@ func (b *builtinCastDurationAsTimeSig) evalTime(row []types.Datum) (res types.Ti
 	res, err = res.RoundFrac(b.tp.Decimal)
 	return res, false, errors.Trace(err)
 }
+
+type castType byte
```
Comment (shenli) [inline expression/builtin_cast.go]: Add comment for this.

**Comment by shenli** [comment]
@hanfei1991 PTAL

File: expression/builtin_cast.go (hunk 1 above)
Comment (XuHuaiyu) [inline expression/builtin_cast.go]: No, `mysql.TypeDuration` is regarded as `types.ClassString`
We first check the `TypeClass` of an Expression,
if it's `types.ClassString`, then we check whether it is `mysql.TypeDuration`.

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991 @shenli

File: expression/builtin_cast.go [hunk 4]
```diff
@@ -558,3 +817,110 @@ func (b *builtinCastDurationAsTimeSig) evalTime(row []types.Datum) (res types.Ti
 	res, err = res.RoundFrac(b.tp.Decimal)
 	return res, false, errors.Trace(err)
 }
+
+func buildCastFunction(expr Expression, tp *types.FieldType, ctx context.Context) (*ScalarFunction, error) {
+	var fc functionClass
+	switch tp.ToClass() {
+	case types.ClassInt:
```
Comment (hanfei1991) [inline expression/builtin_cast.go]: What if the expr is int type? At that time, we needn't add a cast.

File: util/types/etc.go [hunk 5]
```diff
@@ -62,6 +62,12 @@ func IsTypeFractionable(tp byte) bool {
 	return tp == mysql.TypeDatetime || tp == mysql.TypeDuration || tp == mysql.TypeTimestamp
 }
 
+// IsTypeTime returns a boolean indicating
+// whether the tp is time type like datetime, date or timestamp.
+func IsTypeTime(tp byte) bool {
```
Comment (hanfei1991) [inline util/types/etc.go]: Is this function only used by expression pakcage? If so , we needn't put it in util package.

Comment (XuHuaiyu) [inline util/types/etc.go]: There are many other IsTypeXXX functions in this file, 
I'm not sure whether it'll be called in other packages,
so I put it here.

File: expression/builtin_cast.go (hunk 4 above)
Comment (XuHuaiyu) [inline expression/builtin_cast.go]: we check it in line854,
buildCastFunction is only responsible for build the `cast` ScalarFunction,
thus can be reused in  NewCastFunction.

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**